File: leetcode/minimum_path_sum.py
# Given a m x n grid filled with non-negative numbers, find a path from top 
# left to bottom right which minimizes the sum of all numbers along its path.
# Time Complexity: O(mn)
# Space Complexity: O(1)

class Solution:
    # @param {integer[][]} grid
    # @return {integer}
    def minPathSum(self, grid):
        for i in xrange(len(grid)):
            for j in xrange(len(grid[i])):
                if i > 0 and j > 0: 
                    grid[i][j] += min(grid[i - 1][j], grid[i][j - 1])
                elif i > 0:
                    grid[i][j] += grid[i - 1][j]
                elif j > 0:
                    grid[i][j] += grid[i][j - 1]
        return grid[-1][-1]

if __name__ == "__main__":
    solution = Solution()
    print solution.minPathSum([ [0,0,0], [0,1,0], [0,0,0] ])

File: leetcode/minimum_path_sum.js
/* Given a m x n grid filled with non-negative numbers, find a path from top 
 * left to bottom right which minimizes the sum of all numbers along its path.
 * Time Complexity: O(mn)
 * Space Complexity: O(1)
 */
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function(grid) {
    for (var i = 0; i < grid.length; ++i) {
        for (var j = 0; j < grid[i].length; ++j) {
            if (i > 0 && j > 0)
                grid[i][j] += Math.min(grid[i - 1][j], grid[i][j - 1])
            else if (i > 0)
                grid[i][j] += grid[i - 1][j]
            else if (j > 0)
                grid[i][j] += grid[i][j - 1]
        }
    }
    return grid[grid.length - 1][grid[0].length - 1]
};
